packages/flutter/lib/src/painting/debug_overflow_indicator.dart

```diff
@@ -0,0 +1,392 @@
+        'before putting it in the $runtimeType, or using a scrollable '
+        'container, like a ListView, for example.';
+  }
+
+  /// Returns a [RelativeRect] that describes the overflow conditions
+  /// calculated.
+  ///
+  /// Values in the result will be positive if there is an overflow on
+  /// that side.
+  RelativeRect get overflow {
+    _overflow ??= new RelativeRect.fromRect(
+      _overflowContainerRect,
+      _overflowChildRect,
+    );
+    return _overflow;
+  }
+
+  RelativeRect _overflow;
+
+  /// The [Rect] of the child of the container to be checked for
+  /// overflow.
+  ///
+  /// Both [overflowChildRect] and [overflowContainerRect] must be set before
+  /// the overflow will be calculated correctly.
+  Rect get overflowChildRect => _overflowChildRect;
+  set overflowChildRect(Rect overflowChildRect) {
+    if (overflowChildRect == _overflowChildRect)
+      return;
+    _overflowChildRect = overflowChildRect;
+    _markOverflowNeedsUpdate();
+  }
+
+  Rect _overflowChildRect = Rect.zero;
+
+  /// The [Rect] of the container that is to be checked for overflow.
+  ///
+  /// Both [overflowChildRect] and [overflowContainerRect] must be set before
+  /// the overflow will be calculated correctly.
+  Rect get overflowContainerRect => _overflowContainerRect;
+  set overflowContainerRect(Rect overflowContainerRect) {
+    if (overflowContainerRect == _overflowContainerRect)
+      return;
+    _overflowContainerRect = overflowContainerRect;
+    _markOverflowNeedsUpdate();
+  }
+
+  Rect _overflowContainerRect = Rect.zero;
+
+  void _markOverflowNeedsUpdate() {
+    _overflow = null;
+  }
+
+  /// Returns true if the [overflowContainerRect] has been found to overflow.
+  bool get isOverflowing =>
+      overflow.left > 0.0 ||
+      overflow.right > 0.0 ||
+      overflow.top > 0.0 ||
+      overflow.bottom > 0.0;
+
+  String _formatPixels(double value) {
+    String pixels = 'zero';
+    if (value <= 0.0)
+      return pixels;
+    if (value > 10.0) {
+      pixels = value.toStringAsFixed(0);
+    } else if (value > 1.0) {
+      pixels = value.toStringAsFixed(1);
+    } else {
+      pixels = value.toStringAsPrecision(3);
+    }
+    return pixels;
+  }
+
+  List<_OverflowRegionData> _calculateOverflowRegions() {
+    final List<_OverflowRegionData> regions = <_OverflowRegionData>[];
+    if (overflow.left > 0.0) {
+      final Rect markerRect = new Rect.fromLTWH(
+        0.0,
+        0.0,
+        _overflowContainerRect.width * _kMarkerSize,
+        _overflowContainerRect.height,
+      );
+      String label;
+      final String pixels = _formatPixels(overflow.left);
+      if (textDirection != null) {
+        switch (textDirection) {
+          case TextDirection.ltr:
+            label = 'BEGIN OVERFLOWED BY $pixels PIXELS';
+            break;
+          case TextDirection.rtl:
+            label = 'END OVERFLOWED BY $pixels PIXELS';
+            break;
+        }
```

Comment:
I'm not sure it's worth trying to use the begin/end terms here. The presence of textDirection is not a reliable indicator that the user is describing things in those terms at the widgets layer, and "begin" and "end" is probably confusing terminology anyway (also, we use start and end, rather than begin and end).

Reply:
if you remove this logic, you can simplify a lot of the code and remove textDirection from the abstract class' API.

Reply:
OK, will do. I thought might help with debugging to know if the overflow was happening at the "end" and to show you which side the container considered to be the end, but I guess it's not a huge help, really.

Reply:
Yeah if anything it might actually be confusing, because in many cases the developer will just be using left and right (or centering, in the UnconstrainedBox case, and seeing it overflow on both sides, in which case it isn't helpful either).